src/main/java/st/ratpack/auth/CachingTokenValidator.java

```diff
@@ -22,38 +22,25 @@ public CachingTokenValidator(TokenValidator upstreamValidator) {
 			.maximumSize(10000L)
 			.expireAfterWrite(5L, TimeUnit.MINUTES)
 			//			.executor(Execution.current().getEventLoop())  Don't do this it makes Ratpack hang.
-			.build();
+			.build(this::loadCache);
 	}
 
 	@Override
 	public Promise<Optional<OAuthToken>> validate(String token) {
-		Promise<Optional<OAuthToken>> promiseOAuthToken = cache.getIfPresent(token);
-
-		if (promiseOAuthToken == null) {
-			//Cache miss use the upstream validator
-			logger.trace("Cache MISS: " + token);
-			promiseOAuthToken = upstreamValidator.validate(token);
-
-			//Make sure we only call the validate on the upstream once
-			promiseOAuthToken = promiseOAuthToken.cache();
-
-			promiseOAuthToken.then(optionalOAuth -> {
-				logger.trace("PUTTING: " + optionalOAuth);
-				cache.put(token, Promise.value(optionalOAuth));
-			});
-
-			//TODO The wiretap is cleaner but doesn't currently seem to work due to upstream bug
-			//Wiretap the promise so we can cache any successful results
-			//			promiseOAuthToken.wiretap(optionalResult -> {
-			//				logger.trace("Wiretap called");
-			//				if (optionalResult.isSuccess()) {
-			//					logger.trace("Caching promise value");
-			//					cache.put(token, Promise.value(optionalResult.getValue()));
-			//				}
-			//			});
-		} else {
-			logger.trace("CACHE HIT: " + token);
-		}
+		return cache.get(token);
+	}
+
+	private Promise<Optional<OAuthToken>> loadCache(String token) {
+		logger.trace("Cache MISS: {}", token);
+		//Make sure we only call the validate on the upstream once
+		Promise<Optional<OAuthToken>>
+				promiseOAuthToken = upstreamValidator.validate(token).cache();
+
+		promiseOAuthToken
+				.onError(e -> cache.invalidate(token))
+				.map(o -> o.orElse(null))
+				.onNull(() -> cache.invalidate(token))
```

Comment:
Lets say that the `upstreamValidator` returned a completed and exceptional promise. Will the `onError` be deferred to happen sometime after the `loadCache` method completes or will it run immediately?

The reason I ask is because `ConcurrentHashMap` can fail if a map operation recurses to mutate the map. This can live lock in JDK8 (or fail fast when patched) and `CompleteableFuture` has the above mentioned optimization. It requires a little more work to handle if Ratpack's Promise does the same.